examples/drivencavity.py

```diff
@@ -1,60 +1,141 @@
+# form shows that constant pressure shifts do not affect the momentum balance.
+# Both issues are solved by arbitrarily removing one of the basis functions.
+#
+# The normal velocity components at the wall are strongly constrained. The
+# tangential components are either strongly or weakly constrained depending on
+# the `strongbc` parameter. Since the tangential components at the top are
+# incompatible with the normal components at the left and right boundary, the
+# constraints are constructed in two steps to avoid Gibbs type oscillations,
+# and to make sure that the non-penetrating condition takes precedence.
+#
+# Depending on the `compatible` parameter, the script uses either a Taylor-Hood
+# (False) or a Raviart-Thomas (True) discretization. In case of TH, the system
+# is consistently modified by adding ∫_Ω .5 u_i v_i ∇_j(u_j) to yield the skew-
+# symmetric advective term ∫_Ω .5 (v_i ∇_j(u_i) - u_i ∇_j(v_i)) u_j. In case of
+# RT, the discretization guarantees a pointwise divergence-free velocity field,
+# and skew symmetry is implied.
 
 from nutils import mesh, function, solver, export, cli, testing
 from nutils.expression_v2 import Namespace
+import treelog as log
 import numpy
-import treelog
 
 
-def main(nelems: int, etype: str, degree: int, reynolds: float):
+def main(nelems: int, etype: str, degree: int, reynolds: float, compatible: bool, strongbc: bool):
     '''
     Driven cavity benchmark problem.
 
     .. arguments::
 
-       nelems [12]
+       nelems [32]
          Number of elements along edge.
        etype [square]
          Element type (square/triangle/mixed).
        degree [2]
          Polynomial degree for velocity; the pressure space is one degree less.
        reynolds [1000]
          Reynolds number, taking the domain size as characteristic length.
+       strongbc [no]
+         Use strong boundary constraints
+       compatible [no]
+         Use compatible spaces and weakly imposed boundary conditions; requires
+         etype=square and strongbc=no
     '''
 
+    if compatible and (strongbc or etype != 'square'):
+        raise Exception(f'compatible mode requires square elements and weak boundary conditions')
+
     domain, geom = mesh.unitsquare(nelems, etype)
 
     ns = Namespace()
     ns.δ = function.eye(domain.ndims)
+    ns.ε = function.levicivita(2)
     ns.Σ = function.ones([domain.ndims])
     ns.Re = reynolds
+    ns.uwall = numpy.stack([domain.boundary.indicator('top'), 0])
     ns.x = geom
     ns.define_for('x', gradient='∇', normal='n', jacobians=('dV', 'dS'))
-    ns.add_field(('u', 'v'), domain.basis('std', degree=degree), shape=(domain.ndims,))
-    ns.add_field(('p', 'q'), domain.basis('std', degree=degree-1))
+    if not compatible:
+        ns.add_field(('u', 'v'), domain.basis('std', degree=degree), shape=(domain.ndims,))
+        ns.add_field(('p', 'q'), domain.basis('std', degree=degree-1)[1:])
+        ns.add_field('ψ', domain.basis('std', degree=2)[1:])
```

Comment:
The degree of the stream function doesn't follow the degree of the other fields. Is that correct?

Reply:
for non-compatible spaces this seems to be the best setting regardless of velocity degree